File: Graphs/Graphs_dfs.cpp
// Graphs Adjacency List implementation for Generic Data
// Program Author : Abhisek Kumar Gupta
#include<bits/stdc++.h>
using namespace std;

template<typename T>
class Graph{
    map<T, list<T> > adjList;
    public:
        Graph(){

        }
        void addEdge(T u, T v, bool bidir = true){
            adjList[u].push_back(v);
            if(bidir){
                adjList[v].push_back(u);
            }
        }
        void printAdjList(){
            for(auto obj : adjList){
                cout << obj.first <<  "->";
                for(auto element : obj.second){
                    cout << element << ",";
                }
                cout << endl;
            }
        }
        void dfsHelper(int node, map<int, bool> &visited){
            visited[node] = true;
            cout << node << "->";
            for(int neighbour : adjList[node]){
                if(!visited[neighbour]){
                    dfsHelper(neighbour, visited);
                }
            }
        }
        void dfs(int src){
            map<int, bool> visited;
            dfsHelper(src, visited);
        }
};
int main(){
    Graph<int> g;
    g.addEdge(0, 1);
    g.addEdge(0, 4);
    g.addEdge(1, 4);
    g.addEdge(1, 3);
    g.addEdge(1, 2);
    g.addEdge(2, 3);
    g.addEdge(4, 3);
    g.printAdjList();
    g.dfs(0);
    return 0;
}

File: Graphs/Graphs_dfs.py
from collections import defaultdict

class Graph:
    def __init__(self):
        self.graph = defaultdict(list)
    
    def insertEdge(self,v1,v2):
        self.graph[v1].append(v2)
    
    def dfsSub(self,v,visited):

        visited.add(v)
        print(v,end=" ")

        for neighbour in self.graph[v]:
            if neighbour not in visited:
                self.dfsSub(neighbour,visited)

    def dfs(self,v):

        visited = set()
        self.dfsSub(v,visited)

g=Graph()
g.insertEdge(1,2)
g.insertEdge(2,1)
g.insertEdge(3,4)
g.insertEdge(4,5)
g.insertEdge(5,6)
g.insertEdge(2,6)

g.dfs(2)
        
